### DISCUSSION TITLE
Avoid repeated entering of database password in CLI

### DISCUSSION THREAD

**Discussion by mbeko** [keepassxreboot/keepassxc#8231] [Q&A] [answered]
I want to use passwords from the database in Bash scripts and I tried to use `keepassxc-cli show` for that.

However, I have to enter the password every time that I call the command. This goes against the idea of automation with scripts and it seems inconsistent when the KeePassXC GUI offers various options to keep the database unlocked. In fact, I usually keep the database unlocked in the background via the GUI, so it seems strange to be forced to unlock it again when accessed via CLI.

How can I avoid that? I had the following ideas:

* Use CLI commands to explicitly lock and unlock the database for following commands
* Keep the database unlocked after typing the password for the first CLI command
* Access the database that has been unlocked via GUI from the CLI

Is any of this possible? The only other option that I could think of is accessing a database by keyfile only. But I'd prefer to stick to a password to avoid the trouble of securing the keyfile against unauthorised access or loss.

I'm aware of `keepassxc-cli open`, but this starts an interactive session, so it doesn't seem suitable for scripts.

**Comment by droidmonkey** [comment, marked as answer]
`keepassxc-cli open` is your friend. Otherwise the use of the cli is transactional and won't ever be different.

**Comment by mbeko** [comment]
How so? Have you read my last sentence?

For example:
I want to run a script that has to get a password, execute Bash commands, get another password, execute more commands.

How does `keepassxc-cli open` help me here? Can it execute other commands than those of `keepassxc-cli`?
The man page doesn't say much, I'd be happy to get some details how it is supposed to work for that case.

**Reply by droidmonkey** [reply]
You have two choices: 
1. Ask for the credentials for your database at the start of your script and store them in a variable. Use that variable to unlock the database on each command.
2. Use the open command once and read/write to it's STDOUT/STDIN

**Reply by mbeko** [reply]
Thanks for the suggestions, I'll try them out!

Sorry, this is the first time I use GitHub discussions and only now I realised that my reply via e-mail didn't open a thread on your suggested answer. Now I can't mark your thread reply as the answer and marking my post seems wrong. So I'll just mark your first post.

**Reply by mbeko** [reply]
It seems that the second suggestion doesn't work.

Piping to the stdin of the open command causes it to terminate immediately, no matter if the correct password is provided:

```shell
$ echo 'correctpw' | keepassxc-cli open /tmp/Passwords.kdbx
Enter password to unlock /tmp/Passwords.kdbx: 
Test> %
$ echo 'wrongpw' | keepassxc-cli open /tmp/Passwords.kdbx
Enter password to unlock /tmp/Passwords.kdbx: 
Error while reading the database: Invalid credentials were provided, please try again.
If this reoccurs, then your database file may be corrupt. (HMAC mismatch)
> %
$   
```
(The percent sign is added by my shell to mark the end of stream)

> Otherwise the use of the cli is transactional and won't ever be different.

Does this still hold after finding out that the open command doesn't behave as assumed for non-interactive use?

I'd like to know if I should file a bug report for the open command or a feature request for making the non-interactive CLI commands work with unlocked databases or if any of these will be rejected immediately.

**Reply by droidmonkey** [reply]
What version of keepassxc are you using? I thought we fixed that issue with the open command.  @louib

**Reply by mbeko** [reply]
I'm using the KeePassXC package from the Ubuntu repository, it's version 2.6.6.

I'll be busy today, but I'll check as soon as possible if the problem is gone with version 2.7.1.

**Reply by louib** [reply]
@droidmonkey I was not aware of the `open` issue. @mbeko please open an issue if you can reproduce with 2.7.1

> making the non-interactive CLI commands work with unlocked databases

There might already be an issue for that, but otherwise you can create one. I'm guessing we would require confirmation from the user before allowing the CLI to interact with a running GUI instance.

**Reply by mbeko** [reply]
Thanks for the reminder, I had forgotten to check with 2.7.1. I did so now and I can reproduce the behaviour, so I'll open an issue for that:

```shell
echo 'correctpw' | ./KeePassXC-2.7.1-x86_64.AppImage cli open /tmp/Passwords.kdbx
Enter password to unlock /tmp/Passwords.kdbx: 
Passwords> %
```

I'll also check if an issue about using unlocked databases from the CLI exists. I think user confirmation would be fine as long as it doesn't have to be done for every single CLI command.

**Reply by mbeko** [reply]
I created #8969 for the `open` issue.

Regarding using databases unlocked via GUI from the CLI, I found #1483. Especially [this comment describes what I had in mind](https://github.com/keepassxreboot/keepassxc/issues/1483#issuecomment-579408718). However, the issue has been closed with the statement that [this won't be implemented](https://github.com/keepassxreboot/keepassxc/issues/1483#issuecomment-1074379859).

**Comment by kud** [comment]
Hey. I'm not sure to follow if there's a way to handle the opening of the database from the GUI and uses the cli as much as I want without re-opening the database with a password?

As long as the database is opened via the GUI, I'd like to be able to discuss with the database via any protocol I want.

I feel it's something which is done for Browsers for instance.

Any idea?

**Reply by droidmonkey** [reply]
Not possible and we decided to not support that.  There are third party solutions implemented in various languages (eg Python) that support using the browser protocol.

**Reply by kud** [reply]
Oh, that's interesting. Thanks. I'll have a look. 😌

**Reply by kud** [reply]
By any chance, do you know if there's a client in node.js? I only find ones in Rust or Python.
